### PR TITLE
Batch transformers

### LINKED ISSUE #72: Implement new _transform methods in BatchTransformerMixins
Now that we have added the functionality thanks to #69 we should implement batch processing for transformers that can gain from it. This includes a fair chunk of the API and Google transformers.

### PR DESCRIPTION
**Pull request by qmac**
Still a WIP, need to sweep over other potential batch transformers.

Resolves #72. Added core logic in `transformers.py` to perform batch transformations within pliers.

### CODE REVIEW

**Comment by tyarkoni** [comment]
Is this ready to merge, or still WIP?

**Comment by qmac** [comment]
Still a WIP

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ dist: trusty
 
 env:
   global:
-    - CONDA_DEPS="pip pytest numpy pillow pandas requests scipy nltk six tqdm seaborn matplotlib pathos tensorflow"
+    - CONDA_DEPS="pip pytest numpy pillow pandas requests scipy nltk six tqdm seaborn matplotlib pathos tensorflow contextlib2"
```
Comment (qmac) [inline .travis.yml]: Added this package to do nested, looping context managers in the Clarifai extractor (see below)

File: pliers/extractors/base.py [hunk 2]
```diff
@@ -40,15 +40,6 @@ def __init__(self, data, stim, extractor, features=None, onsets=None,
...
         self._history = None
-
-        # Some Extractors pass a list of Stims for various reasons
-        if isinstance(stim, (list, tuple)):
-            if not len(set(stim)) == 1:
-                raise ValueError("Multiple Stim instances passed to the 'stim'"
-                                 " argument in ExtractorResult initialization;"
-                                 " each ExtractorResult can only be associated"
-                                 " with a single Stim.")
-            stim = stim[0]
```
Comment (qmac) [inline pliers/extractors/base.py]: I don't think this is necessary

**Comment by qmac** [comment]
This should finally be ready. @tyarkoni take a look when you get the chance.

File: pliers/converters/google.py [hunk 3]
```diff
@@ -25,31 +25,34 @@ def __init__(self, handle_annotations='first', *args, **kwargs):
...
+                annotations = response[self.response_object]
+                # Combine the annotations
+                if self.handle_annotations == 'first':
+                    text = annotations[0]['description']
+                    texts.append(TextStim(text=text, onset=stim.onset,
+                                 duration=stim.duration))
+                elif self.handle_annotations == 'concatenate':
+                    text = ''
+                    for annotation in annotations:
+                        text += annotation['description']
```
Comment (tyarkoni) [inline pliers/converters/google.py]: Didn't test this, but will this preserve whitespace between tokens, or should we add spaces?

File: pliers/extractors/api.py [hunk 4]
```diff
@@ -19,7 +24,7 @@
     pass
 
 
-class IndicoAPIExtractor(Extractor):
+class IndicoAPIExtractor(BatchTransformerMixin, Extractor):
```
Comment (tyarkoni) [inline pliers/extractors/api.py]: Don't know if there are any alternatives, but I'm fine with this for now.

File: pliers/extractors/api.py [hunk 5]
```diff
@@ -31,6 +36,7 @@ class IndicoAPIExtractor(Extractor):
 
     _log_attributes = ('models',)
     _input_type = ()
+    _batch_size = 20
```
Comment (tyarkoni) [inline pliers/extractors/api.py]: Is there an easy way to make this an initialization argument for any class that implements `BatchTransformerMixin`? Can we check `kwargs` for a `batch_size` argument, and overwrite the class-level `_batch_size` value if it's found?

File: pliers/extractors/api.py [hunk 6]
```diff
@@ -160,11 +149,16 @@ def __init__(self, app_id=None, app_secret=None, model=None,
         else:
             self.select_classes = ','.join(select_classes)
 
-    def _extract(self, stim):
-        with stim.get_filename() as filename:
-            with open(filename, 'rb') as f:
-                tags = self.tagger.tag_images(f, select_classes=self.select_classes)
+    def _extract(self, stims):
```
Comment (tyarkoni) [inline pliers/extractors/api.py]: Maybe add an in-line comment or two describing what's happening here

File: pliers/transformers.py [hunk 7]
```diff
@@ -152,13 +150,32 @@ def __hash__(self):
         return hash(self.name + str(dict(zip(self._log_attributes, tr_attrs))))
 
 
-class BatchTransformerMixin(object):
+class BatchTransformerMixin(Transformer):
```
Comment (tyarkoni) [inline pliers/transformers.py]: Might have missed it, but we should probably add a test or two that test this class as directly as possible (e.g., with a dummy test class that inherits from `BatchTransformerMixin`).

**Comment by qmac** [comment]
I believe I've addressed your comments in f5b2584

**Pull request merged**